execution/aggregate/khashaggregate.go

```diff
@@ -205,41 +209,113 @@ func (a *kAggregate) init(ctx context.Context) error {
+	groupsRemaining := len(a.heaps)
+
+	if a.aggregation != parser.LIMITK {
+		for i, sId := range sampleIDs { // BOTTOMK, TOPK
+			h := a.inputToHeap[sId]
+			switch {
+			case h.Len() < k:
+				heap.Push(h, &entry{sId: sId, total: samples[i]})
 
-		case h.compare(h.entries[0].total, samples[i]) || (math.IsNaN(h.entries[0].total) && !math.IsNaN(samples[i])):
-			h.entries[0].sId = sId
-			h.entries[0].total = samples[i]
+			case h.compare(h.entries[0].total, samples[i]) || (math.IsNaN(h.entries[0].total) && !math.IsNaN(samples[i])):
+				h.entries[0].sId = sId
+				h.entries[0].total = samples[i]
 
-			if k > 1 {
-				heap.Fix(h, 0)
+				if k > 1 {
+					heap.Fix(h, 0)
+				}
+			}
+		}
+	} else {
+		if len(histogramIDs) == 0 {
+			for i, sId := range sampleIDs {
+				h := a.inputToHeap[sId]
+				switch {
+				case h.Len() < k:
+					heap.Push(h, &entry{sId: sId, total: samples[i]})
+
+					if h.Len() == k && a.aggregation == parser.LIMITK {
+						groupsRemaining--
+					}
+
+					if groupsRemaining == 0 {
+						break
+					}
+				}
+			}
+		} else {
+			// Mixed type histograms or native histograms only
+			histogramIndex := 0
+			sampleIndex := 0
+
+			for histogramIndex < len(histogramIDs) || sampleIndex < len(sampleIDs) {
+				var currentID uint64
+
+				// Pick the relatively first id in the sample
+				if histogramIndex < len(histogramIDs) && sampleIndex < len(sampleIDs) {
+					currentID = uint64(min(sampleIDs[sampleIndex], histogramIDs[histogramIndex]))
+				} else if histogramIndex < len(histogramIDs) {
+					// Only histograms left
+					currentID = histogramIDs[histogramIndex]
+				} else {
+					// Only samples left
+					currentID = sampleIDs[sampleIndex]
+				}
+
+				h := a.inputToHeap[currentID]
+
+				if h.Len() < k {
+					if histogramIndex < len(histogramIDs) && histogramIDs[histogramIndex] == currentID {
+						heap.Push(h, &entry{histId: currentID, histogramSample: histograms[histogramIndex]})
+						histogramIndex++
+					}
+
+					if sampleIndex < len(sampleIDs) && sampleIDs[sampleIndex] == currentID {
+						heap.Push(h, &entry{sId: currentID, total: samples[sampleIndex]})
+						sampleIndex++
+					}
+
+					if h.Len() == k {
+						groupsRemaining--
+					}
+
+					if groupsRemaining == 0 {
+						break
+					}
+				} else {
+					if histogramIndex < len(histogramIDs) && histogramIDs[histogramIndex] == currentID {
+						histogramIndex++
+					}
+					if sampleIndex < len(sampleIDs) && sampleIDs[sampleIndex] == currentID {
+						sampleIndex++
+					}
+				}
 			}
 		}
 	}
 
 	s := a.vectorPool.GetStepVector(t)
 	for _, h := range a.heaps {
-		// The heap keeps the lowest value on top, so reverse it.
-		if len(h.entries) > 1 {
-			sort.Sort(sort.Reverse(h))
-		}
-
 		for _, e := range h.entries {
-			s.AppendSample(a.vectorPool, e.sId, e.total)
+			if a.aggregation != parser.LIMITK || e.histogramSample == nil {
```

Comment:
I suggest doing this differently to make it more future-proof:

```
if e.histogramSample != nil {
  s.AppendHistogram(...)
} else {
  s.AppendSample(...)
}
```

What kind of aggregation it is and what kind of sample types are supported by them are handled one layer above.